File: solutions/scala/zebra-puzzle/2/src/main/scala/ZebraPuzzle.scala
object ZebraPuzzle {

  sealed trait Resident
  case object Englishman extends Resident
  case object Spaniard extends Resident
  case object Ukrainian extends Resident
  case object Norwegian extends Resident
  case object Japanese extends Resident

  sealed trait Color
  case object Red extends Color
  case object Green extends Color
  case object Ivory extends Color
  case object Yellow extends Color
  case object Blue extends Color

  sealed trait Pet
  case object Dog extends Pet
  case object Snail extends Pet
  case object Fox extends Pet
  case object Horse extends Pet
  case object Zebra extends Pet

  sealed trait Beverage
  case object Coffee extends Beverage
  case object Tea extends Beverage
  case object Milk extends Beverage
  case object OrangeJuice extends Beverage
  case object Water extends Beverage

  sealed trait Cigarette
  case object OldGold extends Cigarette
  case object Kool extends Cigarette
  case object Chesterfield extends Cigarette
  case object LuckyStrike extends Cigarette
  case object Parliament extends Cigarette

  case class Solution(waterDrinker: Resident, zebraOwner: Resident)

  val colors: Array[Color] = Array(Red, Green, Ivory, Yellow, Blue)
  val nations: Array[Resident] =
    Array(Englishman, Spaniard, Ukrainian, Norwegian, Japanese)
  val pets: Array[Pet] = Array(Dog, Snail, Fox, Horse, Zebra)
  val beverages: Array[Beverage] = Array(Coffee, Tea, Milk, OrangeJuice, Water)
  val cigarettes: Array[Cigarette] =
    Array(OldGold, Kool, Chesterfield, LuckyStrike, Parliament)

  def rightOf[A, B](
      array1: Array[A],
      value1: A,
      array2: Array[B],
      value2: B
  ) =
    array1.indexOf(value1) - array2.indexOf(value2) == 1

  def nextTo[A, B](
      array1: Array[A],
      value1: A,
      array2: Array[B],
      value2: B
  ) =
    (array1.indexOf(value1) - array2.indexOf(value2)).abs == 1

  def sameHouse[A, B](
      array1: Array[A],
      value1: A,
      array2: Array[B],
      value2: B
  ) =
    array1.indexOf(value1) == array2.indexOf(value2)

  def colorRules(colorPerm: Array[Color]) =
    rightOf(colorPerm, Green, colorPerm, Ivory)

  def nationRules(colorPerm: Array[Color], nationPerm: Array[Resident]) =
    sameHouse(colorPerm, Red, nationPerm, Englishman) &&
      nextTo(colorPerm, Blue, nationPerm, Norwegian) &&
      nationPerm(0) == Norwegian

  def petRules(
      colorPerm: Array[Color],
      nationPerm: Array[Resident],
      petPerm: Array[Pet]
  ) =
    sameHouse(nationPerm, Englishman, petPerm, Dog)

  def beverageRules(
      colorPerm: Array[Color],
      nationPerm: Array[Resident],
      petPerm: Array[Pet],
      beveragePerm: Array[Beverage]
  ) =
    sameHouse(beveragePerm, Coffee, colorPerm, Green) &&
      sameHouse(nationPerm, Ukrainian, beveragePerm, Tea) &&
      beveragePerm(2) == Milk

  def cigaretteRules(
      colorPerm: Array[Color],
      nationPerm: Array[Resident],
      petPerm: Array[Pet],
      beveragePerm: Array[Beverage],
      cigarettePerm: Array[Cigarette]
  ) =
    sameHouse(cigarettePerm, OldGold, petPerm, Snail) &&
      sameHouse(cigarettePerm, Kool, colorPerm, Yellow) &&
      nextTo(cigarettePerm, Chesterfield, petPerm, Fox) &&
      nextTo(cigarettePerm, Kool, petPerm, Horse) &&
      sameHouse(cigarettePerm, LuckyStrike, beveragePerm, OrangeJuice) &&
      sameHouse(nationPerm, Japanese, cigarettePerm, Parliament)

  lazy val solutions = for {
    colorPerm <- colors.permutations if colorRules(colorPerm)
    nationPerm <- nations.permutations if nationRules(colorPerm, nationPerm)
    petPerm <- pets.permutations if petRules(colorPerm, nationPerm, petPerm)
    beveragePerm <- beverages.permutations
    if beverageRules(colorPerm, nationPerm, petPerm, beveragePerm)
    cigarettePerm <- cigarettes.permutations if cigaretteRules(
      colorPerm,
      nationPerm,
      petPerm,
      beveragePerm,
      cigarettePerm
    )
  } yield Solution(
    waterDrinker = nationPerm(beveragePerm.indexOf(Water)),
    zebraOwner = nationPerm(petPerm.indexOf(Zebra))
  )

  lazy val solve: Solution = solutions.toList.head
}

File: solutions/prolog/zebra-puzzle/2/zebra_puzzle.pl
color(red).
color(green).
color(yellow).
color(ivory).
color(blue).

country(englishman).
country(spaniard).
country(ukrainian).
country(norwegian).
country(japanese).

pet(dog).
pet(snails).
pet(fox).
pet(horse).
pet(zebra).

beverage(coffee).
beverage(tea).
beverage(milk).
beverage(orange_juice).
beverage(water).

cigarette(old_gold).
cigarette(kool).
cigarette(chesterfield).
cigarette(lucky_strike).
cigarette(parliament).

house([Color, Country, Pet, Beverage, Cigarette]) :-
    color(Color),
    country(Country),
    pet(Pet),
    beverage(Beverage),
    cigarette(Cigarette).

right_of(H1, H2, [H1, H2 | _]).
right_of(H1, H2, [_ | Rest]) :-
    right_of(H1, H2, Rest).

next_to(H1, H2, Houses) :-
    right_of(H1, H2, Houses).
next_to(H1, H2, Houses) :-
    right_of(H2, H1, Houses).

different([X | Xs], [Y | Ys]) :-
    X \= Y,
    different(Xs, Ys).
different([], []).

all_different(Houses) :-
    foreach(select(House1, Houses, Rest), foreach(member(House2, Rest), different(House1, House2))).

solution(Houses) :-
    length(Houses, 5), % Statement 1
    member([red, englishman | _], Houses), % 2
    member([_, spaniard, dog | _], Houses), % 3
    member([green, _, _, coffee | _], Houses), % 4
    member([_, ukrainian, _, tea | _], Houses), % 5
    right_of([ivory | _], [green | _], Houses), % 6
    member([_, _, snails, _, old_gold], Houses), % 7
    member([yellow, _, _, _, kool], Houses), % 8
    [_, _, [_, _, _, milk, _], _, _] = Houses, % 9
    [[_, norwegian, _, _, _] | _] = Houses, % 10
    next_to([_, _, _, _, chesterfield], [_, _, fox | _], Houses), % 11
    next_to([_, _, _, _, kool], [_, _, horse | _], Houses), % 12
    member([_, _, _, orange_juice, lucky_strike ], Houses), % 13
    member([_, japanese, _, _, parliament], Houses), % 14
    next_to([_, norwegian | _], [blue | _], Houses), % 15 */
    maplist(house, Houses),
    all_different(Houses).

zebra_owner(Owner) :-
    solution(Houses),
    member([_, Owner, zebra, _, _], Houses).


water_drinker(Drinker) :-
    solution(Houses),
    member([_, Drinker, _, water, _], Houses).
